### PR TITLE
[SPARK-24137][K8s] Mount local directories as empty dir volumes.

### PR DESCRIPTION
**Pull request by mccheah**
## What changes were proposed in this pull request?

Drastically improves performance and won't cause Spark applications to fail because they write too much data to the Docker image's specific file system. The file system's directories that back emptydir volumes are generally larger and more performant.

## How was this patch tested?

Has been in use via the prototype version of Kubernetes support, but lost in the transition to here.

### CODE REVIEW

**Comment by mccheah** [comment]
@foxish @liyinan926 please take a look, thanks!

**Comment by andrusha** [comment]
Seems like it addresses similar problem to https://github.com/apache/spark/pull/21095. It might be worth investigating how to unify both.

**Comment by mccheah** [comment]
Also #21260 currently only supports hostPath and PVCs but you definitely want emptyDir for isolation (though that looks like a trivial enough change).

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 1]
```diff
@@ -455,7 +455,8 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
   private[spark] def validateSettings() {
     if (contains("spark.local.dir")) {
       val msg = "In Spark 1.0 and later spark.local.dir will be overridden by the value set by " +
-        "the cluster manager (via SPARK_LOCAL_DIRS in mesos/standalone and LOCAL_DIRS in YARN)."
+        "the cluster manager (via SPARK_LOCAL_DIRS in mesos/standalone/kubernetes and LOCAL_DIRS" +
```
Comment (foxish) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: oops, I deleted a comment here accidentally. @rxin said that we could remove this warning about Spark 1.0.

**Comment by erikerlandson** [comment]
I agree with @mcheah that the potential code reuse is small. Keeping this as a separate pod construction step, decoupled from the user-exposed step, is cleaner.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  private val resolvedLocalDirs = Option(conf.sparkConf.getenv("SPARK_LOCAL_DIRS"))
+    .orElse(conf.getOption("spark.local.dir"))
+    .getOrElse(defaultLocalDir)
+    .split(",")
+
+  override def configurePod(pod: SparkPod): SparkPod = {
+    val localDirVolumes = resolvedLocalDirs
+      .zipWithIndex
+      .map {
+        case (localDir, index) =>
```
Comment (liyinan926) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: I think the convention is to put `case` on the same line as `map {`.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          new VolumeBuilder()
+            .withName(s"spark-local-dir-${index + 1}-${Paths.get(localDir).getFileName.toString}")
```
Comment (liyinan926) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Do you really need to include the actual path in the volume name? I think `spark-local-dir-${index + 1}` is sufficient.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            .withNewEmptyDir()
+            .endEmptyDir()
+            .build()
+      }
+    val localDirVolumeMounts = localDirVolumes
+      .zip(resolvedLocalDirs)
+      .map {
+        case (localDirVolume, localDirPath) =>
```
Comment (liyinan926) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Ditto.

**Comment by mccheah** [comment]
@rxin @liyinan926 @foxish addressed comments.

**Review by liyinan926 — APPROVED** [review]
LGTM.

**Comment by mccheah** [comment]
Requesting sign off and merge from @foxish

**Comment by foxish** [comment]
LGTM. Merging to master. 
Thanks @mccheah

**Comment by liyinan926** [comment]
@foxish @mccheah should we also merge this to branch-2.3? I think we should target this for 2.3.1.

**Comment by foxish** [comment]
@mccheah, wdyt? I just haven't heard from any users here of 2.3 - if you think it's useful for 2.3.1 and low risk, then please feel free to propose a cherrypick.

**Comment by mccheah** [comment]
I think we can afford to hold off here.

**Comment by mccheah** [comment]
What would make this difficult to backport is the fact that this patch was built on top of the big refactor PR that only went in after 2.3. So we'd need to rewrite this with the old architecture which is a non-trivial effort.

**Comment by foxish** [comment]
SG. @liyinan926, let's revisit this if we hear from 2.3 users.

**Comment by liyinan926** [comment]
Makes sense to me.

**Pull request closed without merging**